sound/soc/intel/boards/sof_maxim_common.c

```diff
@@ -326,17 +381,28 @@ const struct snd_soc_ops max_98390_ops = {
 };
 EXPORT_SYMBOL_NS(max_98390_ops, SND_SOC_INTEL_SOF_MAXIM_COMMON);
 
-void max_98390_set_codec_conf(struct snd_soc_card *card, int ch)
+void max_98390_dai_link(struct snd_soc_dai_link *link)
+{
+	link->num_codecs = get_maxim_num_codec();
+	if (link->num_codecs == 4)
+		link->codecs = max_98390_4spk_components;
+
+	link->codecs = max_98390_components;
+	link->init = max_98390_spk_codec_init;
+	link->ops = &max_98390_ops;
+}
+EXPORT_SYMBOL_NS(max_98390_dai_link, SND_SOC_INTEL_SOF_MAXIM_COMMON);
+
+void sof_max98390_codec_conf(struct snd_soc_card *card)
 {
-	if (ch == ARRAY_SIZE(max_98390_4spk_codec_conf)) {
+	if (get_maxim_num_codec() == 4) {
 		card->codec_conf = max_98390_4spk_codec_conf;
 		card->num_configs = ARRAY_SIZE(max_98390_4spk_codec_conf);
-	} else {
-		card->codec_conf = max_98390_codec_conf;
-		card->num_configs = ARRAY_SIZE(max_98390_codec_conf);
 	}
+	card->codec_conf = max_98390_codec_conf;
+	card->num_configs = ARRAY_SIZE(max_98390_codec_conf);
```

Comment:
card->codec_conf will be always max_98390_codec_conf and never be max_98390_4spk_codec_conf.